pkg/collector/collector.go

```diff
@@ -0,0 +1,534 @@
+	if h.WorkerStartTime != nil {
+		ch <- prometheus.MustNewConstMetric(
+			uc.coreUpgradeMetrics.upgradeState,
+			prometheus.GaugeValue,
+			float64(h.WorkerStartTime.Time.Unix()),
+			upgradeConfig.Spec.Desired.Version,
+			valueWorkersStarted,
+		)
+	}
+}
+
+func (uc *UpgradeCollector) collectWorkerCompleteTime(upgradeConfig *upgradev1alpha1.UpgradeConfig, h *upgradev1alpha1.UpgradeHistory, ch chan<- prometheus.Metric) {
+	if h.WorkerCompleteTime != nil {
+		ch <- prometheus.MustNewConstMetric(
+			uc.coreUpgradeMetrics.upgradeState,
+			prometheus.GaugeValue,
+			float64(h.WorkerCompleteTime.Time.Unix()),
+			upgradeConfig.Spec.Desired.Version,
+			valueWorkersCompleted,
+		)
+	}
+}
+
+func (uc *UpgradeCollector) collectWorkerTimeout(upgradeConfig *upgradev1alpha1.UpgradeConfig, h *upgradev1alpha1.UpgradeHistory, ch chan<- prometheus.Metric) {
+
+	workerTimeout := upgradeConfig.Status.History[0].WorkerTimeout
+
+	if workerTimeout {
+		ch <- prometheus.MustNewConstMetric(
+			uc.coreUpgradeMetrics.workerTimeout,
+			prometheus.GaugeValue,
+			float64(1),
+			upgradeConfig.ObjectMeta.Name,
+			upgradeConfig.Spec.Desired.Version,
+		)
+		return
+	}
+
+	ch <- prometheus.MustNewConstMetric(
+		uc.coreUpgradeMetrics.workerTimeout,
+		prometheus.GaugeValue,
+		float64(0),
+		upgradeConfig.ObjectMeta.Name,
+		upgradeConfig.Spec.Desired.Version,
+	)
+}
+
+func (uc *UpgradeCollector) collectNodeDrainFailed(upgradeConfig *upgradev1alpha1.UpgradeConfig, h *upgradev1alpha1.UpgradeHistory, ch chan<- prometheus.Metric) {
+
+	drainFailed := upgradeConfig.Status.History[0].NodeDrain.Failed
+
+	if drainFailed {
+		ch <- prometheus.MustNewConstMetric(
+			uc.coreUpgradeMetrics.nodeDrainFailed,
+			prometheus.GaugeValue,
+			float64(1),
+			upgradeConfig.Status.History[0].NodeDrain.Name,
+		)
+		return
+	}
+
+	ch <- prometheus.MustNewConstMetric(
+		uc.coreUpgradeMetrics.nodeDrainFailed,
+		prometheus.GaugeValue,
+		float64(0),
+		upgradeConfig.Status.History[0].NodeDrain.Name,
+	)
+}
+func (uc *UpgradeCollector) collectCompleteTime(upgradeConfig *upgradev1alpha1.UpgradeConfig, h *upgradev1alpha1.UpgradeHistory, ch chan<- prometheus.Metric) {
+	if h.CompleteTime != nil {
+		ch <- prometheus.MustNewConstMetric(
+			uc.coreUpgradeMetrics.upgradeState,
+			prometheus.GaugeValue,
+			float64(h.CompleteTime.Time.Unix()),
+			upgradeConfig.Spec.Desired.Version,
+			valueCompleted,
+		)
+	}
+}
+
+func (uc *UpgradeCollector) collectControlPlaneTimeout(upgradeConfig *upgradev1alpha1.UpgradeConfig, h *v1.UpdateHistory, ch chan<- prometheus.Metric) {
+
+	controlPlaneTimeout := upgradeConfig.Status.History[0].ControlPlaneTimeout
+
+	if controlPlaneTimeout {
+		ch <- prometheus.MustNewConstMetric(
+			uc.coreUpgradeMetrics.controlPlaneTimeout,
+			prometheus.GaugeValue,
+			float64(1),
+			upgradeConfig.ObjectMeta.Name,
+			valueControlPlaneStarted,
```

Comment:
Is there a reasoning for including the "control_plane_started" label in the metric related to timeout? (same goes for its inclusion below)